### PR TITLE
Change originator node: added tests

### PR DESCRIPTION
**Pull request by irynamatveieva**
## Pull Request description

Added tests for the change originator node.

PE PR: https://github.com/thingsboard/thingsboard-pe/pull/2751

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [x] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java [hunk 1]
```diff
@@ -80,74 +108,247 @@ public void originatorCanBeChangedToCustomerId() {
...
         ArgumentCaptor<TbMsg> msgCaptor = ArgumentCaptor.forClass(TbMsg.class);
         ArgumentCaptor<EntityId> originatorCaptor = ArgumentCaptor.forClass(EntityId.class);
-        verify(ctx).transformMsgOriginator(msgCaptor.capture(), originatorCaptor.capture());
+        then(ctxMock).should().transformMsgOriginator(msgCaptor.capture(), originatorCaptor.capture());
 
-        assertEquals(customerId, originatorCaptor.getValue());
+        assertThat(originatorCaptor.getValue()).isEqualTo(customerId);
+    }
+
+    public void init() throws TbNodeException {
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java:125]: 1. why public?
2. why is this method called only in one place, maybe we can remove it?

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java [hunk 2]
```diff
@@ -80,74 +108,247 @@ public void originatorCanBeChangedToCustomerId() {
...
         RuleNodeId ruleNodeId = new RuleNodeId(Uuids.timeBased());
 
-        TbMsg msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, assetId, TbMsgMetaData.EMPTY, TbMsgDataType.JSON, TbMsg.EMPTY_JSON_OBJECT, ruleChainId, ruleNodeId);
-
-        when(ctx.getAssetService()).thenReturn(assetService);
-        when(assetService.findAssetByIdAsync(any(),eq( assetId))).thenReturn(Futures.immediateFuture(asset));
+        TbMsg msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, assetId, TbMsgMetaData.EMPTY, TbMsgDataType.JSON,TbMsg.EMPTY_JSON_OBJECT, ruleChainId, ruleNodeId);
 
-        node.onMsg(ctx, msg);
+        given(ctxMock.getDbCallbackExecutor()).willReturn(dbExecutor);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java:113]: duplicate logic. Already mocked in the init method.

**Review by ShvaykaD — CHANGES_REQUESTED** [review]

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java [hunk 3]
```diff
@@ -20,134 +20,343 @@
...
+        given(ctxMock.getAssetService()).willReturn(assetServiceMock);
+        given(ctxMock.getTenantId()).willReturn(TENANT_ID);
+        given(assetServiceMock.findAssetByTenantIdAndName(any(TenantId.class), any(String.class))).willReturn(null);
+
+        node.init(ctxMock, new TbNodeConfiguration(JacksonUtil.valueToTree(config)));
+        node.onMsg(ctxMock, msg);
+
+        ArgumentCaptor<Throwable> throwable = ArgumentCaptor.forClass(Throwable.class);
+        then(ctxMock).should().tellFailure(eq(msg), throwable.capture());
+        assertThat(throwable.getValue()).isInstanceOf(IllegalStateException.class).hasMessage("Failed to found ASSET  entity by name: 'test-asset'!");
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java:359]:
```suggestion
        assertThat(throwable.getValue()).isInstanceOf(IllegalStateException.class).hasMessage("Failed to found asset with name 'test-asset'!");
```

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java [hunk 4]
```diff
@@ -20,134 +20,343 @@
...
-        Asset asset = new Asset();
-        asset.setCustomerId(customerId);
+    public void givenEntitySourceAndUnexpectedEntityType_whenInit_thenThrowsException() {
+        config.setOriginatorSource(ENTITY_SOURCE);
+        config.setEntityType(EntityType.TENANT.name());
+        config.setEntityNamePattern("tenant-A");
+
+        assertThatThrownBy(() -> node.init(ctxMock, new TbNodeConfiguration(JacksonUtil.valueToTree(config))))
+                .isInstanceOf(IllegalStateException.class)
+                .hasMessage("Unexpected entity type TENANT");
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java:204]:
```suggestion
                .hasMessage("Unexpected entity type 'TENANT'.");
```

**Review by ShvaykaD — CHANGES_REQUESTED** [review]

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java [hunk 5]
```diff
@@ -110,26 +114,26 @@ private ListenableFuture<? extends EntityId> getNewOriginator(TbContext ctx, TbM
     private void validateConfig(TbChangeOriginatorNodeConfiguration conf) {
         HashSet<String> knownSources = Sets.newHashSet(CUSTOMER_SOURCE, TENANT_SOURCE, RELATED_SOURCE, ALARM_ORIGINATOR_SOURCE, ENTITY_SOURCE);
         if (!knownSources.contains(conf.getOriginatorSource())) {
-            log.error("Unsupported source [{}] for TbChangeOriginatorNode", conf.getOriginatorSource());
-            throw new IllegalArgumentException("Unsupported source TbChangeOriginatorNode" + conf.getOriginatorSource());
+            log.error("Unsupported source type '[{}]'! Only {} types are allowed.", conf.getOriginatorSource(), supportedOriginatorSourcesStr);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java:117]:
```suggestion
            log.error("Unsupported source type '{}'! Only {} types are allowed.", conf.getOriginatorSource(), supportedOriginatorSourcesStr);
```

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java [hunk 6]
```diff
@@ -110,26 +114,26 @@ private ListenableFuture<? extends EntityId> getNewOriginator(TbContext ctx, TbM
...
+                log.error("Relations query should be specified if 'Related entity' source is selected.");
+                throw new IllegalArgumentException("Relations query should be specified if 'Related entity' source is selected.");
             }
         }
 
         if (conf.getOriginatorSource().equals(ENTITY_SOURCE)) {
             if (conf.getEntityType() == null) {
-                log.error("Entity type not specified for [{}]", ENTITY_SOURCE);
-                throw new IllegalArgumentException("Wrong config for [{}] in TbChangeOriginatorNode!" + ENTITY_SOURCE);
+                log.error("Entity type should be specified if '[{}]' source is selected.", ENTITY_SOURCE);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java:131]:
```suggestion
                log.error("Entity type should be specified if '{}' source is selected.", ENTITY_SOURCE);
```

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java [hunk 7]
```diff
@@ -110,26 +114,26 @@ private ListenableFuture<? extends EntityId> getNewOriginator(TbContext ctx, TbM
...
+                throw new IllegalArgumentException("Entity type should be specified if 'Entity by name pattern' source is selected.");
             }
             if (StringUtils.isEmpty(conf.getEntityNamePattern())) {
-                log.error("EntityNamePattern not specified for type [{}]", conf.getEntityType());
-                throw new IllegalArgumentException("Wrong config for [{}] in TbChangeOriginatorNode!" + ENTITY_SOURCE);
+                log.error("Name pattern should be specified if '[{}]' source is selected.", ENTITY_SOURCE);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java:135]:
```suggestion
                log.error("Name pattern should be specified if '{}' source is selected.", ENTITY_SOURCE);
```

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java [hunk 8]
```diff
@@ -20,134 +20,343 @@
...
-        init();
+        node = new TbChangeOriginatorNode();
+        config = new TbChangeOriginatorNodeConfiguration().defaultConfiguration();
     }
 
     @Test
-    public void originatorCanBeChangedToCustomerId() {
-        AssetId assetId = new AssetId(Uuids.timeBased());
-        CustomerId customerId = new CustomerId(Uuids.timeBased());
+    public void newChainCanBeStarted() throws TbNodeException {
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java:119]: please remove this outdated test.

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java [hunk 9]
```diff
@@ -20,134 +20,343 @@
...
+        device.setCustomerId(CUSTOMER_ID);
+
+        TbMsg msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, DEVICE_ID, TbMsgMetaData.EMPTY, TbMsg.EMPTY_JSON_OBJECT);
+        TbMsg expectedMsg = TbMsg.transformMsgOriginator(msg, CUSTOMER_ID);
+
+        given(ctxMock.getDbCallbackExecutor()).willReturn(dbExecutor);
+        given(ctxMock.getDeviceService()).willReturn(deviceServiceMock);
+        given(ctxMock.getTenantId()).willReturn(TENANT_ID);
+        given(deviceServiceMock.findDeviceById(any(TenantId.class), any(DeviceId.class))).willReturn(device);
+        given(ctxMock.transformMsgOriginator(any(TbMsg.class), any(EntityId.class))).willReturn(expectedMsg);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNodeTest.java:219]: please add a verification to check what will be passed to the transformMsgOriginator method. Add this in other places as well.

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesByNameAndTypeLoader.java [hunk 10]
```diff
@@ -53,7 +53,7 @@ public static EntityId findEntityId(TbContext ctx, EntityType entityType, String
                 throw new IllegalStateException("Unexpected entity type " + entityType.name());
         }
         if (targetEntity == null) {
-            throw new IllegalStateException("Failed to found " + entityType.name() + "  entity by name: '" + entityName + "'!");
+            throw new IllegalStateException("Failed to found " + entityType.name().toLowerCase() + " with name '" + entityName + "'!");
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesByNameAndTypeLoader.java:56]:
```suggestion
            throw new IllegalStateException("Failed to found " + entityType.getNormalName().toLowerCase() + " with name '" + entityName + "'!");
```

**Review by ShvaykaD — CHANGES_REQUESTED** [review]

**Review by ShvaykaD — APPROVED** [review]

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java [hunk 11]
```diff
@@ -110,26 +114,26 @@ private ListenableFuture<? extends EntityId> getNewOriginator(TbContext ctx, TbM
     private void validateConfig(TbChangeOriginatorNodeConfiguration conf) {
         HashSet<String> knownSources = Sets.newHashSet(CUSTOMER_SOURCE, TENANT_SOURCE, RELATED_SOURCE, ALARM_ORIGINATOR_SOURCE, ENTITY_SOURCE);
```
Comment (ViacheslavKlimov) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java:115]: Please extract to constant

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java [hunk 12]
```diff
@@ -110,26 +114,26 @@ private ListenableFuture<? extends EntityId> getNewOriginator(TbContext ctx, TbM
...
             if (conf.getEntityType() == null) {
-                log.error("Entity type not specified for [{}]", ENTITY_SOURCE);
-                throw new IllegalArgumentException("Wrong config for [{}] in TbChangeOriginatorNode!" + ENTITY_SOURCE);
+                log.error("Entity type should be specified if '{}' source is selected.", ENTITY_SOURCE);
+                throw new IllegalArgumentException("Entity type should be specified if 'Entity by name pattern' source is selected.");
             }
             if (StringUtils.isEmpty(conf.getEntityNamePattern())) {
-                log.error("EntityNamePattern not specified for type [{}]", conf.getEntityType());
-                throw new IllegalArgumentException("Wrong config for [{}] in TbChangeOriginatorNode!" + ENTITY_SOURCE);
+                log.error("Name pattern should be specified if '{}' source is selected.", ENTITY_SOURCE);
```
Comment (ViacheslavKlimov) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java:135]: Let's change log level to DEBUG

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesByNameAndTypeLoader.java [hunk 13]
```diff
@@ -53,7 +53,7 @@ public static EntityId findEntityId(TbContext ctx, EntityType entityType, String
                 throw new IllegalStateException("Unexpected entity type " + entityType.name());
         }
         if (targetEntity == null) {
-            throw new IllegalStateException("Failed to found " + entityType.name() + "  entity by name: '" + entityName + "'!");
+            throw new IllegalStateException("Failed to found " + entityType.getNormalName().toLowerCase() + " with name '" + entityName + "'!");
```
Comment (ViacheslavKlimov) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesByNameAndTypeLoader.java:56]: Failed to find*

**Review by ViacheslavKlimov — CHANGES_REQUESTED** [review]

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java [hunk 14]
```diff
@@ -112,27 +103,19 @@ private ListenableFuture<? extends EntityId> getNewOriginator(TbContext ctx, TbM
...
-            log.error("Unsupported source type '{}'! Only {} types are allowed.", conf.getOriginatorSource(), supportedOriginatorSourcesStr);
-            throw new IllegalArgumentException("Unsupported source type '" + conf.getOriginatorSource() +
-                    "'! Only " + supportedOriginatorSourcesStr + " types are allowed.");
-        }
-
-        if (conf.getOriginatorSource().equals(RELATED_SOURCE)) {
+        if (conf.getOriginatorSource().equals(RELATED)) {
             if (conf.getRelationsQuery() == null) {
-                log.error("Relations query should be specified if 'Related entity' source is selected.");
+                log.debug("Relations query should be specified if 'Related entity' source is selected.");
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbChangeOriginatorNode.java:108]: merge if statements. In addition, please add the check for null for: conf.getOriginatorSource()

**Review by ShvaykaD — APPROVED** [review]

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/ChangeOriginatorSource.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.rule.engine.transform;
+
+public enum ChangeOriginatorSource {
```
Comment (ViacheslavKlimov) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/ChangeOriginatorSource.java:18]: Let's rename it to OriginatorSource

**Review by ViacheslavKlimov — APPROVED** [review]

**Pull request merged**